File: src/main/java/by/dev/madhead/jarvis/pipeline/JarvisStepExecution.java
package by.dev.madhead.jarvis.pipeline;

import by.dev.madhead.jarvis.Messages;
import by.dev.madhead.jarvis.Jarvis;
import by.dev.madhead.jarvis.creator.EmailCreatorFactory;
import by.dev.madhead.jarvis.util.RecipientParser;
import hudson.AbortException;
import hudson.FilePath;
import hudson.model.TaskListener;
import org.jenkinsci.plugins.workflow.job.WorkflowRun;
import org.jenkinsci.plugins.workflow.steps.StepContext;
import org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution;

import javax.annotation.Nonnull;
import javax.mail.MessagingException;
import java.io.IOException;

public class JarvisStepExecution extends SynchronousNonBlockingStepExecution {

    private final String defaultRecipients;
    private final boolean tlsEnable;

    JarvisStepExecution(@Nonnull StepContext context, String defaultRecipients, boolean tlsEnable) {
        super(context);
        this.defaultRecipients = defaultRecipients;
        this.tlsEnable = tlsEnable;
    }

    @Override
    protected Object run() throws Exception {
        WorkflowRun run = getContext().get(WorkflowRun.class);
        if (run == null) {
            throw new AbortException(Messages.jarvis_hudson_AbortException_workflowRunRequired());
        }
        TaskListener listener = getContext().get(TaskListener.class);
        if (listener == null) {
            throw new AbortException(Messages.jarvis_hudson_AbortException_taskListenerRequired());
        }
        FilePath workspace = getContext().get(FilePath.class);
        if (workspace == null) {
            throw new AbortException(Messages.jarvis_hudson_AbortException_workspaceRequired());
        }

        Jarvis jarvis = new Jarvis(tlsEnable);
        try {
            jarvis.sendMail(EmailCreatorFactory.getCreator(run, listener, workspace).create(),
                    RecipientParser.createDefaultAddressesSet(run, defaultRecipients));
        } catch (IOException | MessagingException e) {
            throw new AbortException(e.getMessage());
        }
        return null;
    }

}

File: src/main/kotlin/by/dev/madhead/jarvis/pipeline/JarvisStepExecution.kt
package by.dev.madhead.jarvis.pipeline

import by.dev.madhead.jarvis.Jarvis
import by.dev.madhead.jarvis.Messages
import by.dev.madhead.jarvis.creator.EmailCreator
import by.dev.madhead.jarvis.util.createDefaultAddressesSet
import hudson.AbortException
import hudson.FilePath
import hudson.model.TaskListener
import org.jenkinsci.plugins.workflow.job.WorkflowRun
import org.jenkinsci.plugins.workflow.steps.StepContext
import org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution

class JarvisStepExecution(
        context: StepContext,
        val defaultRecipients: String?) : SynchronousNonBlockingStepExecution<Void>(context) {

    override fun run(): Void? {
        val run = context.get(WorkflowRun::class.java)
                ?: throw AbortException(Messages.jarvis_hudson_AbortException_workflowRunRequired())
        val listener = context.get(TaskListener::class.java)
                ?: throw AbortException(Messages.jarvis_hudson_AbortException_taskListenerRequired())
        try {
            Jarvis().sendMail(
                    email = EmailCreator(run, listener,
                            context.get(FilePath::class.java)
                                    ?: throw AbortException(Messages.jarvis_hudson_AbortException_workspaceRequired())
                    ).create(),
                    defaultRecipientsAddresses = createDefaultAddressesSet(run, defaultRecipients),
                    logger = listener.logger)
            return null
        } catch (exception: Throwable) {
            throw AbortException(exception.message)
        }
    }

}
